Source/WebCore/css/StyleProperties.cpp

```diff
@@ -898,61 +988,50 @@ static bool gridAutoFlowContains(const RefPtr<CSSValue>& autoFlow, CSSValueID id
+    auto autoColumnsIndex = longhandIndex(5, CSSPropertyGridAutoColumns);
+
+    auto& autoColumns = longhandValue(autoColumnsIndex);
+    auto& autoRows = longhandValue(autoRowsIndex);
+    auto& autoFlow = longhandValue(autoFlowIndex);
 
     if (isValueIDIncludingList(autoColumns, CSSValueAuto) && isValueIDIncludingList(autoRows, CSSValueAuto) && isValueIDIncludingList(autoFlow, CSSValueRow))
-        return getGridTemplateValue();
+        return serializeGridTemplate();
 
-    if (!isValueID(getPropertyCSSValue(CSSPropertyGridTemplateAreas), CSSValueNone))
+    if (!isLonghandValueNone(areasIndex))
         return String();
 
-    auto rows = getPropertyCSSValue(CSSPropertyGridTemplateRows);
-    auto columns = getPropertyCSSValue(CSSPropertyGridTemplateColumns);
-    if (!rows || !columns)
-        return String();
+    auto& rows = longhandValue(rowsIndex);
+    auto& columns = longhandValue(columnsIndex);
 
-    StringBuilder result;
+    bool autoFlowContainsDense = gridAutoFlowContains(autoFlow, CSSValueDense);
+    auto dense = autoFlowContainsDense ? " dense" : "";
 
     if (gridAutoFlowContains(autoFlow, CSSValueColumn)) {
         if (!isValueIDIncludingList(autoRows, CSSValueAuto) || !isValueIDIncludingList(columns, CSSValueNone))
             return String();
 
-        result.append(rows->cssText());
-        result.append(" / auto-flow");
-        if (gridAutoFlowContains(autoFlow, CSSValueDense))
-            result.append(" dense");
-
-        if (!isValueIDIncludingList(autoColumns, CSSValueAuto))
-            result.append(' ', autoColumns->cssText());
-
-        return result.toString();
+        if (isValueIDIncludingList(autoColumns, CSSValueAuto))
+            return makeString(serializeLonghandValue(rowsIndex), " / auto-flow", dense);
+        return makeString(serializeLonghandValue(rowsIndex), " / auto-flow", dense, ' ', serializeLonghandValue(autoColumnsIndex));
     }
 
-    if (!gridAutoFlowContains(autoFlow, CSSValueRow) && !gridAutoFlowContains(autoFlow, CSSValueDense))
+    if (!gridAutoFlowContains(autoFlow, CSSValueRow) && !autoFlowContainsDense)
         return String();
-
     if (!isValueIDIncludingList(autoColumns, CSSValueAuto) || !isValueIDIncludingList(rows, CSSValueNone))
         return String();
 
-    result.append("auto-flow");
-    if (gridAutoFlowContains(autoFlow, CSSValueDense))
-        result.append(" dense");
-
-    if (!isValueIDIncludingList(autoRows, CSSValueAuto)) {
-        result.append(" ");
-        result.append(autoRows->cssText());
-    }
-
-    result.append(" / ");
-    result.append(columns->cssText());
-
-    return result.toString();
+    if (isValueIDIncludingList(autoRows, CSSValueAuto))
+        return makeString("auto-flow", dense, " / ", serializeLonghandValue(columnsIndex));
+    return makeString("auto-flow", dense, " ", serializeLonghandValue(autoColumnsIndex), " / ", serializeLonghandValue(columnsIndex));
```

Comment:
`autoRowsIndex`?

Reply:
Oops, I need to find out why tests didn't catch this mistake.

Reply:
I see now. They did. The relevant test just wasn’t in my short list of tests that I am running locally.